File: typescript/src/core/DependencyContainer.ts
import { google } from 'googleapis';
import type { Config } from '../types/index.js';
import { FileOperations } from '../services/FileOperations.js';
import { Logger } from '../services/Logger.js';
import { AuthenticationService } from '../services/AuthenticationService.js';
import { GoogleSheetsService } from '../services/GoogleSheetsService.js';
import { GoogleDriveService } from '../services/GoogleDriveService.js';
import { YouTubeService } from '../services/YouTubeService.js';
import { ProgressTracker } from '../services/ProgressTracker.js';
import { VideoProcessor } from './VideoProcessor.js';
import { YouTubeBulkUploader } from './YouTubeBulkUploader.js';

export class DependencyContainer {
  private fileOps: FileOperations;
  private logger: Logger;
  private authService: AuthenticationService;
  private sheetsService?: GoogleSheetsService;
  private driveService?: GoogleDriveService;
  private youtubeService?: YouTubeService;
  private progressTracker: ProgressTracker;
  private videoProcessor?: VideoProcessor;

  constructor(private readonly config: Config) {
    this.fileOps = new FileOperations();
    this.logger = new Logger(config.logFile ?? './upload_log.txt', this.fileOps);
    this.authService = new AuthenticationService(config, this.fileOps, this.logger);
    this.progressTracker = new ProgressTracker(
      config.progressFile ?? './upload_progress.json',
      this.fileOps
    );
  }

  async createYouTubeBulkUploader(): Promise<YouTubeBulkUploader> {
    // Initialize authentication first
    const authClient = await this.authService.initialize();

    // Set up Google API authentication globally
    google.options({ auth: authClient });

    // Create services with authenticated client
    this.sheetsService = new GoogleSheetsService(authClient);
    this.driveService = new GoogleDriveService(authClient, this.fileOps, this.logger);
    this.youtubeService = new YouTubeService(authClient, this.fileOps);

    // Create video processor
    this.videoProcessor = new VideoProcessor(
      this.driveService,
      this.youtubeService,
      this.fileOps,
      this.config
    );

    // Create and return the main uploader
    return new YouTubeBulkUploader(
      this.authService,
      this.sheetsService,
      this.videoProcessor,
      this.progressTracker,
      this.logger,
      this.config
    );
  }
}

File: python/src/core/dependency_container.py
"""Dependency injection container for assembling services."""

from models import Config
from services import (
    AuthenticationService,
    FileOperations,
    GoogleDriveService,
    GoogleSheetsService,
    Logger,
    ProgressTracker,
    YouTubeService,
)

from .video_processor import VideoProcessor
from .youtube_bulk_uploader import YouTubeBulkUploader


class DependencyContainer:
    """Container for managing and injecting dependencies."""

    def __init__(self, config: Config) -> None:
        """
        Initialize dependency container.

        Args:
            config: Application configuration
        """
        self.config = config
        
        # Initialize basic services
        self.file_operations = FileOperations()
        self.logger = Logger(self.file_operations, config.log_file)
        self.progress_tracker = ProgressTracker(self.file_operations, config.progress_file)
        
        # Authentication service (needs to be initialized before Google services)
        self.auth_service = AuthenticationService(
            config, self.file_operations, self.logger
        )

    def create_youtube_bulk_uploader(self) -> YouTubeBulkUploader:
        """
        Create and return a fully configured YouTubeBulkUploader instance.

        Returns:
            Configured YouTubeBulkUploader ready to process videos
        """
        # Initialize authentication and get credentials
        credentials = self.auth_service.initialize()
        
        # Create Google API services with authenticated credentials
        sheets_service = GoogleSheetsService(credentials)
        drive_service = GoogleDriveService(
            credentials, self.file_operations, self.logger
        )
        youtube_service = YouTubeService(credentials, self.file_operations)
        
        # Create video processor
        video_processor = VideoProcessor(
            drive_service, youtube_service, self.file_operations, self.config
        )
        
        # Create and return the main uploader
        return YouTubeBulkUploader(
            auth_service=self.auth_service,
            sheets_service=sheets_service,
            video_processor=video_processor,
            progress_tracker=self.progress_tracker,
            logger=self.logger,
            config=self.config,
        )